### PR TITLE
COMPAT: sum/prod on all nan will remain nan regardless of bottleneck install

### LINKED ISSUE #9422: API: sum of Series of all NaN should return 0 or NaN ?
### Summary

The question is what the sum of a Series of all NaNs should return (which is equivalent to an empty Series after skipping the NaNs): NaN or 0?

```
In [1]: s = Series([np.nan])                 

In [2]: s.sum(skipna=True)  # skipping NaNs is the default
Out[2]: nan or 0     <---- DISCUSSION POINT

In [3]: s.sum(skipna=False)
Out[3]: nan
```

The reason this is a discussion point has the following cause: the internal nansum implementation of pandas returns NaN. But, when bottleneck is installed, pandas will use bottlenecks implementation of nansum, which returns 0 (for the versions >= 1.0). 
Bottleneck changed the behaviour from returning NaN to returning 0 to model it after numpy's nansum function.

This has the very annoying consequence that depending on whether bottleneck is installed or not (which is only an optional dependency), you get a different behaviour.

So the decision we need to make, is to either:

- adapt pandas internal implementation to return 0, so in all cases 0 is returned for all NaN/empty series.
- workaround bottlenecks behaviour or not use it for nansum, in order to consistently return NaN instead of 0
- choose one of both above as the default, but have an option to switch behaviour

---

Original title: **nansum in bottleneck 1.0 will return 0 for all NaN arrays instead of NaN**

xref https://github.com/kwgoodman/bottleneck/issues/96
xref https://github.com/pydata/pandas/issues/9421
- [ ] Tests are turned off for bottleneck >1.0 (xref https://github.com/pydata/pandas/pull/10986)

This matches a change from numpy 1.8 -> 1.9.

We should address this for pandas 0.16.

Should we work around the new behavior (probably the simplest choice) or change nansum in pandas?

### PR DESCRIPTION
**Pull request by jreback**
xref #15507
closes #9422

### CODE REVIEW

**Comment by jreback** [comment]
cc @shoyer 

after all of the conversation of #9422 dead silence here.

**Comment by jreback** [comment]
any objections to context given discussion in #9422 

@jorisvandenbossche @shoyer

File: pandas/tests/frame/test_analytics.py [hunk 1]
```diff
@@ -1859,14 +1856,14 @@ def test_dataframe_clip(self):
...
     def test_clip_mixed_numeric(self):
-
+        # TODO(jreback)
+        # clip on mixed integer or floats
+        # with integer clippers coerces to float
         df = DataFrame({'A': [1, 2, 3],
                         'B': [1., np.nan, 3.]})
         result = df.clip(1, 2)
-        expected = DataFrame({'A': [1, 2, 2],
+        expected = DataFrame({'A': [1, 2, 2.],
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_analytics.py]: is this related ?

**Review by jorisvandenbossche — COMMENTED** [review]
What is the performance impact of not using bottleneck? From a quick test it seems quite a bit faster than our own implementation. 
But I suppose checking for the case of all NaNs (to then not use bottleneck) will defeat this performance gain?

**Comment by jreback** [comment]
> What is the performance impact of not using bottleneck? From a quick test it seems quite a bit faster than our own implementation.
But I suppose checking for the case of all NaNs (to then not use bottleneck) will defeat this performance gain?

there is some benefit, can be up to 2x faster (in ad-hoc tests). But again, we *do* more so its not entirely fair (e.g. handle proper dtypes, infs and things like that). But for some straightforward stuff its fine.

File: pandas/tests/frame/test_analytics.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/frame/test_analytics.py]: no I think was un-xfailing a test and that's all it needed IIRC.

**Comment by jorisvandenbossche** [comment]
> there is some benefit, can be up to 2x faster (in ad-hoc tests). But again, we do more so its not entirely fair (e.g. handle proper dtypes, infs and things like that)

That is true, but depending on the dtype this extra 'work' is not needed. Eg we could special case integers? As there can never be NaN handling needed. Small benchmark:

```
In [24]: df = pd.DataFrame(np.random.randint(0, 1000, (10000, 10)))

In [27]: pd.options.compute.use_bottleneck = True

In [28]: %timeit df.sum()
241 µs ± 1.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [29]: pd.options.compute.use_bottleneck = False

In [30]: %timeit df.sum()
845 µs ± 4.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

So it is a considerable difference.

**Comment by jreback** [comment]
thats a whopping 3us. If you want to do a PR, go ahead.

**Comment by jorisvandenbossche** [comment]
I can easily make that number bigger by making a bigger frame :-) it's the relative number that is relevant. 
But to the point, I am not familiar with the nanops.py code, but it's not just adding a `is_integer_dtype` check here? https://github.com/jreback/pandas/blob/3c91148b1f5b48fa97e475c63a5adfeb429fbdf6/pandas/core/nanops.py#L166 (there is already a check for object and datetime dtype)

**Comment by chris-b1** [comment]
Not sure it's worth the complexity in implementation, but _could_ optimistically use bottleneck, and only if the identity is returned (0 or 1) do a check for all `nan`

**Comment by jreback** [comment]
note by always using our own routine we also avoid some buggy versions on windows  (#15507) though might be fixed in later versions

**Comment by jreback** [comment]
update with more documentation. @jorisvandenbossche @shoyer @TomAugspurger

**Comment by jreback** [comment]
@jorisvandenbossche @TomAugspurger any more comments

File: doc/source/missing_data.rst [hunk 2]
```diff
@@ -181,6 +181,42 @@ account for missing data. For example:
...
+
+Sum/Prod of Empties/Nans
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning::
+
+   This behavior is now standard as of v0.21.0; previously sum/prod would give different
+   results if the ``bottleneck`` package was installed. See the :ref:`here <whatsnew_0210.api_breaking.bottleneck>`.
+
+If summing a ``DataFrame``, a ``Series`` of all-``NaN``.
```
Comment (TomAugspurger) [inline doc/source/missing_data.rst]: typo: ``DataFrame`` or ``Series``

Comment (TomAugspurger) [reply doc/source/missing_data.rst]: And maybe say what the behavior is. "of all-`NaN`, the return value is `NaN`.

File: doc/source/whatsnew/v0.21.0.txt [hunk 3]
```diff
@@ -12,6 +12,7 @@ Highlights include:
 - Integration with `Apache Parquet <https://parquet.apache.org/>`__, including a new top-level :func:`read_parquet` and :func:`DataFrame.to_parquet` method, see :ref:`here <io.parquet>`.
 - New user-facing :class:`pandas.api.types.CategoricalDtype` for specifying
   categoricals independent of the data, see :ref:`here <whatsnew_0210.enhancements.categorical_dtype>`.
+- The behavior of ``sum`` and ``prod`` on all-NaN Series/DataFrames is now consistent and no longer depends on whether `bottleneck <http://berkeleyanalytics.com/bottleneck>`__ is installed, see :ref:`here <whatsnew_0210.api_breaking.bottleneck>`
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: Maybe state that it always returns `NaN`.

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -411,6 +412,54 @@ Current Behavior
...
 
+.. _whatsnew_0210.api_breaking.bottleneck:
+
+Sum/Prod of all-NaN Series/DataFrames is now consistently NaN
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The behavior of ``sum`` and ``prod`` on all-NaN Series/DataFrames is now consistent and no longer depends on
+whether `bottleneck <http://berkeleyanalytics.com/bottleneck>`__ is installed. (:issue:`9422`, :issue:`15507`).
+
+This now will *always* preserve information. You will get back a ``NaN``, indicating missing values in that Series,
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: "indicating **all** missing values"

File: doc/source/whatsnew/v0.21.0.txt [hunk 5]
```diff
@@ -411,6 +412,54 @@ Current Behavior
...
+or if summing a ``DataFrame``, a ``Series`` of all-``NaN``. See the :ref:`docs <missing_data.numeric_sum>`.
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: I don't follow what these two lines are saying. I think I'd phrase it like

    For empty or all-missing ``Series`` or columns of a ``DataFrame``, these operations now return ``NaN``. See the :ref:...

**Review by TomAugspurger — APPROVED** [review]
Just doc comments.

File: doc/source/missing_data.rst [hunk 6]
```diff
@@ -181,6 +181,42 @@ account for missing data. For example:
...
+
+Summing of an empty ``Series``
+
+.. ipython:: python
+
+   pd.Series([]).sum()
+
+.. warning::
+
+   These behaviors differ from the default in ``numpy`` which does not generally propagate NaNs
```
Comment (shoyer) [inline doc/source/missing_data.rst]: Rather than "which does not generally propagate NaNs" I would say "where an empty sum returns zero"

**Review by shoyer — COMMENTED** [review]
Can we also adjust the docstring for `sum`/`prod`?

Currently, it is:
```
skipna : boolean, default True
    Exclude NA/null values. If an entire row/column is NA, the result
    will be NA
```

Maybe this could become:
```
skipna : boolean, default True
    Exclude NA/null values. If an entire row/column is NA or empty, the result
    will be NA.
``

Actually, can confirm the behavior for summing an empty series with `skipna=False`? e.g.,  `pd.Series([]).sum(skipna=False)`?

**Comment by jreback** [comment]
updated

**Comment by shoyer** [comment]
@jreback I had one comment in my last review that I think got lost with bad markdown formatting:

> Can we confirm the behavior for summing an empty series with `skipna=False`? e.g.,  `pd.Series([]).sum(skipna=False)`?

This should be NaN, not 0 like NumPy.

**Comment by jreback** [comment]
0.20.3 
```
# with bottleneck
In [1]: pd.Series([]).sum(skipna=True)
Out[1]: 0

In [2]: pd.Series([]).sum(skipna=False)
Out[2]: 0

In [3]: pd.Series([np.nan]).sum(skipna=True)
Out[3]: 0.0

In [4]: pd.Series([np.nan]).sum(skipna=False)
Out[4]: nan

# no bottleneck
In [1]: pd.Series([]).sum(skipna=True)
Out[1]: 0

In [2]: pd.Series([]).sum(skipna=False)
Out[2]: 0

In [3]: pd.Series([np.nan]).sum(skipna=True)
Out[3]: nan

In [4]: pd.Series([np.nan]).sum(skipna=False)
Out[4]: nan

```

PR
```
In [1]: pd.Series([]).sum(skipna=True)
Out[1]: nan

In [2]: pd.Series([]).sum(skipna=False)
Out[2]: nan

In [3]: pd.Series([np.nan]).sum(skipna=True)
Out[3]: nan

In [4]: pd.Series([np.nan]).sum(skipna=False)
Out[4]: nan
```

**Comment by shoyer** [comment]
Thanks, that looks right!

**Comment by jreback** [comment]
k, merging on green.
